[JavaScript]Add mentoring.md for Triangle (#1624)
exercism/website-copy: tracks/javascript/exercises/triangle/mentoring.md

but you might also want to teach according to students' understanding but they'll be taught in future core exercise so you don't need to stick to them.

- [Rest parameters][ref-rest-parameters]
- [Destructuring assignment][refs-destructuring-assignment]
- [Array#every][ref-array-every]
- [Array#some][ref-array-some]

Also, this exercise might be the first time for some students to use class, so you gently teach them about JavaScript class if needed.

### Reasonable Solutions

```javascript
export class Triangle {
  constructor(...sides) {
    this.sides = sides;
  }

  isEquilateral() {
    const [a, b, c] = this.sides;
    return this.isValid() && a == b && b == c && c == a;
  }

  isIsosceles() {
    const [a, b, c] = this.sides;
    return this.isValid() && (a == b || b == c || c == a);
  }

  isScalene() {
    const [a, b, c] = this.sides;
    return this.isValid() && a != b && b != c && c != a;
  }

  isValid(){
    const [a, b, c] = this.sides;
    const sidesArePositive = a > 0 && b > 0 && c > 0;
    const validatesTriangleInequality =
      a + b >=c && b + c >= a && c + a >= b;
    return sidesArePositive && validatesTriangleInequality;
  }
}
```

> A wise student would use `.sort` to simplify logics.

Reviewer: ```suggestion
A wise student might use `.sort` to simplify the Triangle Inequality rule logic.
```

Author: Thanks!

> A wise student would would use `.sort` to simplify logics.

Reviewer 2: Double "would" there...

> A wise student would would use `.sort` to simplify logics.

Author: Thanks!

```suggestion
A wise student would use `.sort` to simplify logics.
```

Also, some students add test cases for degenerate triangles and implement a method to check that.

> 

Reviewer: ````suggestion

Students use of higher-level concepts (like # of matching sides) is also a reasonable solution and a worthy discussion point for this exercise.

```javascript
isEquilateral() {
  return this.isValid() && this.matchingSides() === 3
}

matchingSides() { 
  // function to calculate the number of matching sides 
}
```

````

Reviewer: Perhaps something like that.

Author: Thanks! What does it mean `#` in `like # of matching sides`?

Reviewer: # == "number", yes... is that not international?  But yeah "number" would be better.

Author: I see. I've never seen it before.

This is the one of the four side exercises unlocked after finishing Resistor Color. The level of understanding is vary from students. Some work on this exercise right after Resistor Color is done, others submit this exercise after solving multiple core exercise.

so you need to teach at least:

- [Logical Operators][ref-logical]

but you might also want to teach according to students' understanding but they'll be taught in future core exercise so you don't need to stick to them.

- [Rest parameters][ref-rest-parameters]
- [Destructuring assignment][refs-destructuring-assignment]
- [Array#every][ref-array-every]
- [Array#some][ref-array-some]

Also, this exercise might be the first time for some students to use class, so you gently teach them about JavaScript class if needed.

### Reasonable Solutions

```javascript
export class Triangle {
  constructor(...sides) {
    this.sides = sides;
  }

  isEquilateral() {
    const [a, b, c] = this.sides;
    return this.isValid() && a == b && b == c && c == a;
  }

  isIsosceles() {
    const [a, b, c] = this.sides;
    return this.isValid() && (a == b || b == c || c == a);
  }

  isScalene() {
    const [a, b, c] = this.sides;

>     return this.isValid() && a != b && b != c && c != a;

Reviewer 3: should we make these strict comparisons?

Reviewer: I'd say yes.

  }

  isValid(){
    const [a, b, c] = this.sides;
    const sidesArePositive = a > 0 && b > 0 && c > 0;
    const validatesTriangleInequality =
      a + b >=c && b + c >= a && c + a >= b;
    return sidesArePositive && validatesTriangleInequality;
  }
}
```

A wise student might use `.sort` to simplify the Triangle Inequality rule logic.

Also, some students add test cases for degenerate triangles and implement a method to check that.

> Students use of higher-level concepts (like # of matching sides) is also a reasonable solution and a worthy discussion point for this exercise.

Author: ```suggestion
Students use of higher-level concepts (like number of matching sides) is also a reasonable solution and a worthy discussion point for this exercise.
```

> Students use of higher-level concepts (like number of matching sides) is also a reasonable solution and a worthy discussion point for this exercise.

Reviewer 3: Nitpick: I think instead of `higher-level concepts`, a better wording would be `separate functions`.

```
Students may choose to use small separate functions (eg. for number of matching sides). This is also a reasonable solution and worthy of discussion.
```

Reviewer: I think the more important thing is actually the "higher level concepts" part, not breaking them into functions... you could use a higher-level concept in a variable and it would still be a huge improvement over super low-level.  

High level concepts do pair well with functions, but they aren't the same thing.

Reviewer: Pseudo-code:

```ts
isScalene() {
  let matchingSides = this.sides.wouldBeNiceIfThisWasBuildIn.orChainAble()
  return this.valid() && matchingSides === 3;
```

Now I do think a function would be nice to wrap that but I'd give huge points for just using the concept - which can then be easily refactored into a function.

```javascript
isEquilateral() {
  return this.isValid() && this.matchingSides() === 3
}

matchingSides() { 
  // function to calculate the number of matching sides 
}
```

#### Approvability

> Some students would use `Set` to solve Triangle exercise and they're also okay if using properly.

Reviewer 3: It would be nice to add the following: `We intentionally try to keep our tests free of implementation details and allow students to choose.`
